x-pack/plugin/ml/build.gradle

```diff
@@ -10,24 +12,40 @@ esplugin {
   extendedPlugins = ['x-pack-autoscaling', 'lang-painless']
 }
 
+def localRepo = providers.systemProperty('build.ml_cpp.repo').orNull
 
 repositories {
-  exclusiveContent {
-    forRepository {
-      ivy {
-        name "ml-cpp"
-        url providers.systemProperty('build.ml_cpp.repo').orElse('https://prelert-artifacts.s3.amazonaws.com').get()
-        metadataSources {
-          // no repository metadata, look directly for the artifact
-          artifact()
-        }
-        patternLayout {
-          artifact "maven/org/elasticsearch/ml/ml-cpp/[revision]/[module]-[revision](-[classifier]).[ext]"
-        }
+  if (localRepo) {
+    ivy {
+      name "ml-cpp"
+      url localRepo
+      metadataSources {
+        // no repository metadata, look directly for the artifact
+        artifact()
+      }
+      patternLayout {
+        artifact "maven/org/elasticsearch/ml/ml-cpp/[revision]/[module]-[revision](-[classifier]).[ext]"
       }
+      content { includeGroup 'org.elasticsearch.ml' }
     }
-    filter {
-      includeGroup 'org.elasticsearch.ml'
+  } else {
+    ivy {
+      name "ml-cpp"
+      url "https://artifacts-snapshot.elastic.co/"
+      metadataSources {
+        // no repository metadata, look directly for the artifact
+        artifact()
+      }
+      patternLayout {
+        if (VersionProperties.isElasticsearchSnapshot()) {
+          artifact '/ml-cpp/[revision]/downloads/ml-cpp/[module]-[revision]-[classifier].[ext]'
+        } else {
+          // When building locally we always use snapshot artifacts even if passing `-Dbuild.snapshot=false`.
+          // Release builds are always done with a local repo.
+          artifact '/ml-cpp/[revision]-SNAPSHOT/downloads/ml-cpp/[module]-[revision]-SNAPSHOT-[classifier].[ext]'
```

Comment:
@droberts195 Can you confirm this is _effectively_ what we were doing before? It seems the ML build was also publishing a non-snapshot version or were those artifacts actual build candidates?

Reply:
The main side effect of this change will be that the "release tests" builds in elasticsearch-ci will now run with snapshot builds of ML.  That is unlikely to make any difference in reality as I cannot think of any time we've made release builds of the ML C++ different to snapshot builds (apart from the version number).  We've never had C++ functionality behind feature flags that's only accessible in snapshot builds for example.

> It seems the ML build was also publishing a non-snapshot version or were those artifacts actual build candidates?

A single ML build was either snapshot or non-snapshot.  But we make sure both are available for any particular version.  The non-snapshot builds are effectively build candidates.  If an emergency release were instigated then the most recent one of those would get incorporated into a unified build candidate.

So I don't think it really matters that much.  What you have at the moment should be fine or changing it to use non-snapshot ML builds in elasticsearch-ci "release tests" would also be fine.  I think the same is probably true for Beats - they appear to regularly publish non-snapshot builds for each branch as part of the DRA process.

Reply:
Thanks David. I'll push forward with this then. I think everyone's cadence for releasing staging builds is potentially different with DRA. I wasn't sure folks produced release builds for any given version prior to FF so I wanted to err on the side of caution and stick with snapshots, since they are most likely to be available immediately after a version bump.